src/engine.cc

```diff
@@ -193,11 +194,20 @@ void EngineController::NewGame() {
   cache_.Clear();
   search_.reset();
   tree_.reset();
+  current_position_.reset();
   UpdateNetwork();
 }
 
 void EngineController::SetPosition(const std::string& fen,
                                    const std::vector<std::string>& moves_str) {
+  current_position_ = CurrentPosition { fen, moves_str };
+
+  SharedLock lock(busy_mutex_);
```

Comment:
It's better to keep that lock in the beginning of the function.
It's not for synchronization (where we need to make it as short as possible), but only for `isready` command. When it's locked, `isready` waits before responding `readyok`.